inc/az_ulib_action_api.h

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+#ifndef AZ_ULIB_ACTION_API_H
+#define AZ_ULIB_ACTION_API_H
+
+#include "azure_macro_utils/macro_utils.h"
+#include "az_ulib_base.h"
+#include "ulib_result.h"
+#include "umock_c/umock_c_prod.h"
+
+#ifdef __cplusplus
+#include <stdint.h>
+extern "C" {
+#include <cstdint>
+#endif /* __cplusplus */
+
+/**
+ * @file    az_ulib_action_api.h
+ *
+ * @brief   Actions are properties, methods and events that can be handled by an interface.
+ */
+
+/**
+ * @brief Enumerator that defines the type of the action.
+ */
+MU_DEFINE_ENUM(
+    AZ_ULIB_ACTION_TYPE,
+    AZ_ULIB_ACTION_TYPE_PROPERTY      = 0x00, /**<Read and write property */
+    AZ_ULIB_ACTION_TYPE_METHOD        = 0x01, /**<Synchronous command that can be invoked by other
+                                              modules in the system */
+    AZ_ULIB_ACTION_TYPE_METHOD_ASYNC  = 0x02, /**<Asynchronous task that can be invoked by other
+                                              modules in the system */
+    AZ_ULIB_ACTION_TYPE_EVENT         = 0x03  /**<Event that other modules in the system can 
+                                              subscribe to be notified */
+);
+
+#define AZ_ULIB_ACTION_TYPE_BITS 0x03
+
+/**
+ * @brief       Token that uniquely identifies the action.
+ *
+ * The action token is the way that the action's caller associates the call to its answer. It can
+ * have any value that is meaningful for the caller.
+ */
+typedef void *az_ulib_action_token;
+
+/**
+ * @brief       Handle that uniquely identifies the action instance.
+ */
+typedef uint16_t az_ulib_action_index;
+
+/**
+ * @brief       Callback prototype for action result.
+ */
+typedef void (*az_ulib_action_result_callback)(
+    const az_ulib_action_token action_token,
+    AZ_ULIB_RESULT result,
+    const void *const model_out);
+
+/**
+ * @brief       Event prototype.
+ */
+typedef void (*az_ulib_action_event)(const void *const model_out);
+
+/**
+ * @brief       Cancellation token prototype.
+ */
+typedef AZ_ULIB_RESULT (*az_ulib_action_cancellation_callback)(
+    const az_ulib_action_token action_token);
+
+/**
+ * @brief       IPC synchronous command signature.
+ */
+typedef AZ_ULIB_RESULT (*az_ulib_action_method)(const void *const model_in, const void *model_out);
+
+/**
+ * @brief       IPC asynchronous task signature.
+ */
+typedef AZ_ULIB_RESULT (*az_ulib_action_method_async)(
+    const void *const model_in,
+    az_ulib_action_result_callback callback,
+    const az_ulib_action_token action_token,
+    az_ulib_action_cancellation_callback *cancel);
```

Comment:
i think these typedef'd functions should have some doxygen function documentation. For ustream release callback for example [here with the generated docs](https://azure.github.io/azure-ulib-c/ustream__base_8h.html#ae39e80e866043231e3535a6270110255) with [this inline in the header](https://github.com/Azure/azure-ulib-c/blob/3c102e56de7d50b8d2630f6acc980a19ca8242a4/inc/ustream_base.h#L368) documentation.

Reply:
I'll do now for `az_ulib_action_method` and for the others when I implement it.

---
In reply to: [339129368](https://github.com/Azure/azure-ulib-c/pull/40#discussion_r339129368) [](ancestors = 339129368)